cumulus/pallets/parachain-system/src/validate_block/implementation.rs

```diff
@@ -179,59 +144,189 @@ where
 			.replace_implementation(host_storage_proof_size),
 	);
 
-	run_with_externalities_and_recorder::<B, _, _>(&backend, &mut recorder, || {
-		let relay_chain_proof = crate::RelayChainStateProof::new(
-			PSC::SelfParaId::get(),
-			inherent_data.validation_data.relay_parent_storage_root,
-			inherent_data.relay_chain_state.clone(),
-		)
-		.expect("Invalid relay chain state proof");
+	let block_data = codec::decode_from_bytes::<ParachainBlockData<B>>(block_data)
+		.expect("Invalid parachain block data");
 
-		#[allow(deprecated)]
-		let res = CI::check_inherents(&block, &relay_chain_proof);
+	let mut parent_header =
+		codec::decode_from_bytes::<B::Header>(parachain_head.clone()).expect("Invalid parent head");
 
-		if !res.ok() {
-			if log::log_enabled!(log::Level::Error) {
-				res.into_errors().for_each(|e| {
-					log::error!("Checking inherent with identifier `{:?}` failed", e.0)
-				});
-			}
+	let (blocks, proof) = block_data.into_inner();
+
+	assert_eq!(
+		*blocks
+			.first()
+			.expect("BlockData should have at least one block")
+			.header()
+			.parent_hash(),
+		parent_header.hash(),
+		"Parachain head needs to be the parent of the first block"
+	);
+
+	let mut processed_downward_messages = 0;
+	let mut upward_messages = BoundedVec::default();
+	let mut upward_message_signals = Vec::<Vec<_>>::new();
+	let mut horizontal_messages = BoundedVec::default();
+	let mut hrmp_watermark = Default::default();
+	let mut head_data = None;
+	let mut new_validation_code = None;
+	let num_blocks = blocks.len();
+
+	// Create the db
+	let db = match proof.to_memory_db(Some(parent_header.state_root())) {
+		Ok((db, _)) => db,
+		Err(_) => panic!("Compact proof decoding failure."),
+	};
 
-			panic!("Checking inherents failed");
-		}
-	});
+	core::mem::drop(proof);
+
+	// We use the same recorder across all blocks. Each node only contributed once to the total size
+	// of the storage proof.
```

Comment:
Reusing it across all blocks makes sense, but it makes me a bit nervous too. 

There is the call to `check_inherents` where people can potentially read whatever they want right? Could stuff end up in the recorder that is not expected during block execution? Before this PR we reset it every time when we called `run_with_externalities_and_recorder`, so it was guaranteed fresh for block execution. Now its called in a loop, so we need to be really careful.

Reply:
Yeah, let me use a fresh one there as well.